File: leetcode/92-reverse-linked-list-II/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseBetween(
  head: ListNode | null,
  left: number,
  right: number,
): ListNode | null {
  // no need to reverse if there is only one node (or 0)
  if (!head || !head.next) {
    return head;
  }

  // let's find the first node to be reversed
  let before: ListNode | null = null;
  let leftNode: ListNode | null = head;

  for (let i = 0; i < left - 1; i++) {
    if (i == left - 2) {
      before = leftNode;
    }

    leftNode = leftNode.next;
  }

  // console.log(`before val: ${before.val}; leftNode val: ${leftNode.val};`);

  // now we're moving to a classic reverse linked list problem
  let prev: ListNode | null = leftNode;
  let curr: ListNode | null = leftNode.next;

  for (let i = left; i < right; i++) {
    const nextNode: ListNode | null = curr.next;
    curr.next = prev;
    prev = curr;
    curr = nextNode;
  }

  // console.log(`After iteration: ${curr.val}, prev: ${prev.val};`);

  // after reversing, update the link from the first node before
  // the "left" and the first node after "right"
  // to correctly point to our reverse (left, right) part of the LL
  if (!before) {
    head = prev;
  } else {
    before.next = prev;
  }

  leftNode.next = curr;

  return head;
}

File: leetcode/92-reverse-linked-list-II/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        // let's find the first node to be reversed
        ListNode* before = nullptr;
        ListNode* left_node = head;

        for (int i = 0; i < left - 1; i++) {
            if (i == left - 2) {
                before = left_node;
            }

            left_node = left_node->next;
        }

        // classic reverse linked list
        ListNode* prev = left_node;
        ListNode* curr = left_node->next;
        for (int i = left; i < right; i++) {
            ListNode* next_node = curr->next;

            curr->next = prev;
            prev = curr;
            curr = next_node;
        }

        // after reversing, update the link from the first node before
        // the "left" and the first node after "right"
        // to correctly point to our reverse (left, right) part of the LL
        if (before == nullptr) {
            head = prev;
        } else {
            before->next = prev;
        }

        left_node->next = curr;

        return head;
    }
};
